### PR TITLE
implement history API endpoint

### PR DESCRIPTION
**Pull request by vladak**
This change adds the API endpoint for history. It supports pagination and history for directories.

### CODE REVIEW

**Comment by vladak** [comment]
I re-ordered the Apiary definitions to be sorted alphabetically (best effort).

File: opengrok-web/src/main/java/org/opengrok/web/api/v1/controller/HistoryController.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Objects;
+import java.util.SortedSet;
+
+@Path("/history")
+public class HistoryController {
+
+    private RuntimeEnvironment env = RuntimeEnvironment.getInstance();
```
Comment (tulinkry) [inline opengrok-web/src/main/java/org/opengrok/web/api/v1/controller/HistoryController.java]: final, static?

File: opengrok-web/src/main/java/org/opengrok/web/api/v1/controller/HistoryController.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(revision, date, author, tags, message, files);
+        }
+    }
+
+    static class HistoryDTO implements JSONable {
+        @JsonProperty
+        private List<HistoryEntryDTO> entries;
```
Comment (tulinkry) [inline opengrok-web/src/main/java/org/opengrok/web/api/v1/controller/HistoryController.java]: final

File: opengrok-web/src/main/java/org/opengrok/web/api/v1/controller/HistoryController.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        @JsonProperty
+        private Date date;
+        @JsonProperty
+        private String author;
+        @JsonProperty
+        private String tags;
+        @JsonProperty
+        private String message;
+        @JsonProperty
+        private SortedSet<String> files;
```
Comment (tulinkry) [inline opengrok-web/src/main/java/org/opengrok/web/api/v1/controller/HistoryController.java]: all can be final

File: opengrok-web/src/main/java/org/opengrok/web/api/v1/controller/HistoryController.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        public int hashCode() {
+            return Objects.hash(entries);
+        }
+    }
+
+    static HistoryDTO getHistoryDTO(List<HistoryEntry> historyEntries) {
+        List<HistoryEntryDTO> entries = new ArrayList<>();
+        for (HistoryEntry entry : historyEntries) {
+            entries.add(new HistoryEntryDTO(entry));
+        }
```
Comment (tulinkry) [inline opengrok-web/src/main/java/org/opengrok/web/api/v1/controller/HistoryController.java]: nb, looks like a task for streaming api

File: opengrok-web/src/test/java/org/opengrok/web/api/v1/controller/HistoryControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    @Test
+    public void testHistoryGet() throws Exception {
+        final String path = "git";
+        Response response = target("history")
+                .queryParam("path", path)
+                .request()
+                .get();
+        HistoryDTO history = response.readEntity(new GenericType<HistoryDTO>() {});
+        History repoHistory = HistoryGuru.getInstance().getHistory(new File(repository.getSourceRoot(), path));
+        assertEquals(history, getHistoryDTO(repoHistory.getHistoryEntries()));
```
Comment (tulinkry) [inline opengrok-web/src/test/java/org/opengrok/web/api/v1/controller/HistoryControllerTest.java]: can you assert something more - contanins some string, minimal length?
it relies on the serialization and when it's broken you wouldn't catch it.

File: apiary.apib [hunk 6]
```diff
@@ -59,11 +67,68 @@ Besides `/suggester` and `/search` endpoints, everything is accessible from `loc
...
-## Authorization framework reload [/configuration/authorization/reload]
+## History [/history]
 
-### reloads authorization framework [POST]
+### get history entries [GET]
 
-+ Request (application/text)
++ Parameters
+  + path (string) - path of file/directory to get history for
+  + withFiles (optional, string) - whether to include list of files
```
Comment (tulinkry) [inline apiary.apib]: string?

**Comment by vladak** [comment]
Authnz checks added, however they can only be used in servlet container per https://stackoverflow.com/a/35868654/11582827 , i.e. the Grizzly test will skip them (and produce bunch of `java.lang.NoSuchMethodException: Could not find a suitable constructor in javax.servlet.http.HttpServletRequest class.` exceptions to the log).

File: apiary.apib (hunk 6 above)
Comment (tulinkry) [reply apiary.apib]: it's boolean not a string

File: opengrok-web/src/main/java/org/opengrok/web/api/v1/controller/HistoryController.java (hunk 3 above)
Comment (vladak) [reply opengrok-web/src/main/java/org/opengrok/web/api/v1/controller/HistoryController.java]: they would have to be initialized in the plain constructor first I guess.

Comment (tulinkry) [reply opengrok-web/src/main/java/org/opengrok/web/api/v1/controller/HistoryController.java]: I get it, it's the serialization issue

File: opengrok-web/src/test/java/org/opengrok/web/api/v1/controller/HistoryControllerTest.java (hunk 5 above)
Comment (vladak) [reply opengrok-web/src/test/java/org/opengrok/web/api/v1/controller/HistoryControllerTest.java]: better now ? :-)

File: opengrok-web/src/test/java/org/opengrok/web/api/v1/controller/HistoryControllerTest.java [hunk 7]
```diff
@@ -105,12 +105,19 @@ public void testHistoryDTOEquals() {
...
+        int start = 2;
         Response response = target("history")
                 .queryParam("path", path)
+                .queryParam("max", size)
+                .queryParam("start", start)
                 .request()
                 .get();
         HistoryDTO history = response.readEntity(new GenericType<HistoryDTO>() {});
+        assertEquals(size, history.getEntries().size());
+        assertEquals("Kryštof Tulinger <krystof.tulinger@oracle.com>", history.getEntries().get(0).getAuthor());
```
Comment (tulinkry) [inline opengrok-web/src/test/java/org/opengrok/web/api/v1/controller/HistoryControllerTest.java]: oh yeah, thank you

**Comment by tulinkry** [comment]
Last question, why would reusing HistoryEntry and making it jsonable be a bad decision? Why you return an object of {entries: []} instead of just plain array [] ?

**Comment by vladak** [comment]
> Last question, why would reusing HistoryEntry and making it jsonable be a bad decision?

It's partly caused by the process - initially I augmented the members of `HistoryEntry` and `History` with `@DTOElement` annotations however I figured out that `DTOUtils` does not work recursively and also has problems with `StringBuffer` (used for `message`) so I went for separate DTO object and finally used `JSONable`. Using dedicated DTO object is less prone to problems with crossing the boundary so I stick to that.

**Comment by tulinkry** [comment]
Ok then. And the second question?

**Comment by tulinkry** [comment]
I'm thinking of it ... when you used object as a returned type instead of an array. You could include additional metadata like "next" or "prev" (page) with the link to the next page and don't set them when it's the last page. Also something like total number would be nice or the input parameters (start and max). I imagine that they can be useful for someone developing against it.

**Comment by vladak** [comment]
> Why do you choose returning an object of `{entries: []}` instead of just plain `array []` ?

For easier testing. It is not strictly necessary. I just pushed a changeset that removes `HistoryDTO` and the endpoint returns a list of history entries.

However, it made me wonder about pagination, e.g. how does one know how many entries in total there are ?

**Comment by vladak** [comment]
Okay, resurrected `HistoryDTO`, added the counters in.

**Comment by tulinkry** [comment]
> However, it made me wonder about pagination, e.g. how does one know how many entries in total there are ?

Just like with github api, one does not know. One tries to get some records and stops when the array becomes empty.

**Comment by tulinkry** [comment]
Let's go

**Comment by vladak** [comment]
> > However, it made me wonder about pagination, e.g. how does one know how many entries in total there are ?
> 
> Just like with github api, one does not know. One tries to get some records and stops when the array becomes empty.

That's a bit hapless I think, especially when the underlying data changes while iterating. Having the total count helps in this regards a bit.

**Comment by tulinkry** [comment]
> That's a bit hapless I think, especially when the underlying data changes while iterating. Having the total count helps in this regards a bit.

I'm not saying it isn't. Just that some giants make it that way.

**Comment by vladak** [comment]
Anyway, thanks for the review !

**Comment by tulinkry** [comment]
You are very welcome.

**Comment by jbaek7023** [comment]
@vladak I just want to say I really appreciate the quick turnaround.

**Pull request merged**